components/content_settings/core/browser/brave_content_settings_pref_provider.cc

```diff
@@ -332,6 +283,76 @@ void BravePrefProvider::MigrateShieldsSettingsV1ToV2() {
   prefs_->SetInteger(kBraveShieldsSettingsVersion, 2);
 }
 
+void BravePrefProvider::MigrateShieldsSettingsV2ToV3() {
+  // Check if migration is needed.
+  if (prefs_->GetInteger(kBraveShieldsSettingsVersion) != 2)
+    return;
+
+  const ContentSettingsPattern& wildcard = ContentSettingsPattern::Wildcard();
+  const ContentSettingsPattern first_party(
+      ContentSettingsPattern::FromString("https://firstParty/*"));
+
+  auto rule_iterator =
+      PrefProvider::GetRuleIterator(ContentSettingsType::BRAVE_COOKIES,
+                                    /*off_the_record*/ false);
+
+  using OldRule = std::pair<ContentSettingsPattern, ContentSettingsPattern>;
+  std::vector<OldRule> old_rules;
+  std::vector<Rule> new_rules;
+
+  // Find rules that can be migrated and create replacement rules for them.
+  while (rule_iterator && rule_iterator->HasNext()) {
+    auto old_rule = rule_iterator->Next();
+    old_rules.emplace_back(old_rule.primary_pattern,
+                           old_rule.secondary_pattern);
+    if (old_rule.primary_pattern == wildcard &&
+        (old_rule.secondary_pattern == wildcard ||
+         old_rule.secondary_pattern == first_party)) {
+      // Remove default rules from BRAVE_COKIES because it's already mapped to
+      // the Chromium prefs.
+      continue;
+    }
+    if (old_rule.secondary_pattern == wildcard && !new_rules.empty() &&
+        new_rules.back().secondary_pattern == old_rule.primary_pattern &&
+        old_rule.value == new_rules.back().value) {
+      // Remove the "first-party" rule because it is a predecessor of a general
+      // rule that we are going to add.
+      new_rules.pop_back();
+    }
+
+    Rule new_rule;
+    new_rule.expiration = old_rule.expiration;
+    new_rule.session_model = old_rule.session_model;
+    new_rule.value = std::move(old_rule.value);
+    // Exchange primary and secondary patterns.
+    new_rule.secondary_pattern = old_rule.primary_pattern;
```

Comment:
instead of fixing the primary/secondary pattern we should just migrate them to the block 3p setting

Reply:
apparently the upstream block 3p setting does not work the way I would expect so we have to keep two settings after all